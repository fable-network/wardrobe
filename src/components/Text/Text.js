import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

export const sizes = ['XS', 'S', 'M', 'L', 'XL', 'XXL', 'XXXL'];

const sizeToFont = {
  XS: 'XSmall',
  S: 'Small',
  M: 'Base',
  L: 'Large',
  XL: 'XLarge',
  XXL: 'XXLarge',
  XXXL: 'XXXLarge',
};

const Text = styled(({ size, bold, renderAs, children, ...otherProps }) =>
  React.createElement(renderAs, otherProps, children)
)`
  font-size: ${p => p.theme[`fontSize${sizeToFont[p.size]}`]};
  line-height: ${p => p.theme[`lineHeight${sizeToFont[p.size]}`]};
  font-weight: ${p => (p.bold ? p.theme.fontWeightBold : p.theme.fontWeightNormal)};
`;

Text.defaultProps = {
  size: 'M',
  bold: false,
  renderAs: 'p',
};

Text.propTypes = {
  size: (props, propName, componentName) => {
    const { size, bold } = props;
    if (!sizes.includes(size)) {
      return new Error(
        `Invalid size: ${size} supplied to ${componentName}. It should be one of ${sizes.join(
          ', '
        )}`
      );
    }
    if (['XL', 'XXL', 'XXXL'].includes(size) && !bold) {
      return new Error(
        `Invalid combination of props supplied to ${componentName}. ${size} can only be used with bold text`
      );
    }
    if (size === 'XS' && bold) {
      return new Error(
        `Invalid combination of props supplied to ${componentName}. ${size} can not be used with bold text`
      );
    }
    return undefined;
  },
  bold: PropTypes.bool,
  renderAs: PropTypes.string,
  children: PropTypes.node,
};

const Heading = ({ size, ...otherProps }) => <Text size={size} bold {...otherProps} />;
Heading.propTypes = {
  size: PropTypes.oneOf(sizes.filter(s => s !== 'XS')),
};

const H1 = props => <Heading renderAs="h1" size="XXXL" {...props} />;
const H2 = props => <Heading renderAs="h2" size="XXL" {...props} />;
const H3 = props => <Heading renderAs="h3" size="XL" {...props} />;
const H4 = props => <Heading renderAs="h4" size="L" {...props} />;
const H5 = props => <Heading renderAs="h5" size="M" {...props} />;
const H6 = props => <Heading renderAs="h6" size="S" {...props} />;

Text.H1 = H1;
Text.H2 = H2;
Text.H3 = H3;
Text.H4 = H4;
Text.H5 = H5;
Text.H6 = H6;

export default Text;
