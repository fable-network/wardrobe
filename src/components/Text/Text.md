The `Text` component saves you from having to individually apply styles for `font-weight`, `font-size`
and `line-height` to components, while using the defined typographic styles. The size and weight of the
text can be defined with the `size` prop and `bold` props which take a t-shirt size (XS - XXXL) and a
boolean. The `renderAs` prop can be used to determine which tag is used for the element (defaults to `p`).

These tags come without margins, so using the [Box](/#/Layout?id=box) component to space them is recommended.

```jsx
<Text bold size="XXXL" renderAs="h1">Some text in the h1 style</Text>
<Text renderAs="div">Some copy in the base size and rendered in a div</Text>
```

### Helper components

There are also the following helper components for headings, which make the text bold and the apply the
correct size. They will also use the correct HTML element (e.g. `<h1>`) though this can also be overridden
by the `renderAs` prop if necessary.

- `<Text.H1>`
- `<Text.H2>`
- `<Text.H3>`
- `<Text.H4>`
- `<Text.H5>`
- `<Text.H6>`

```jsx
<Text.H3>An h3 created using a helper component</Text.H3>
```
