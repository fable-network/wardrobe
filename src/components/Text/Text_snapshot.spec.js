/* eslint-disable */
// This file was automatically generated
import React from 'react';
import { shallow } from 'enzyme';
import Text from './Text';

describe('Text Component', () => {
  describe('snapshots', () => {
    it('should match snapshot', () => {
      const wrapper1 = shallow(
        <Text bold size="XXXL" renderAs="h1">
          Some text in the h1 style
        </Text>
      );
      expect(wrapper1).toMatchSnapshot();
    });
  });
});
