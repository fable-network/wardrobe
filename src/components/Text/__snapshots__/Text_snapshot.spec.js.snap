// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Component snapshots should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Text
    bold={true}
    renderAs="h1"
    size="XXXL"
  >
    Some text in the h1 style
  </Text>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "bold": true,
      "children": "Some text in the h1 style",
      "className": "Text-bbIbrg biTPIt",
      "renderAs": "h1",
      "size": "XXXL",
    },
    "ref": null,
    "rendered": "Some text in the h1 style",
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "bold": true,
        "children": "Some text in the h1 style",
        "className": "Text-bbIbrg biTPIt",
        "renderAs": "h1",
        "size": "XXXL",
      },
      "ref": null,
      "rendered": "Some text in the h1 style",
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
